lib/pbench/test/unit/server/database/test_index_map_db.py

```diff
@@ -37,69 +37,91 @@ def test_create(self, db_session, attach_dataset):
             IndexStream("prefix.run-toc.2023-07", "id6"),
         ]
 
-    def test_merge(self, db_session, attach_dataset):
-        """Test index map merge"""
+    @pytest.mark.parametrize("m1,m2", ((0, 1), (1, 0)))
+    def test_merge(self, db_session, attach_dataset, m1, m2):
+        """Test index map merge
 
-        map1 = {
-            "run-data": {"prefix.run-data.2023-07": ["id1", "id2"]},
-            "run-toc": {
-                "prefix.run-toc.2023-06": ["id3", "id4"],
-                "prefix.run-toc.2023-07": ["id5", "id6"],
+        We merge "a into b" and "b into a" ... the results should be identical.
+        """
+
+        maps = [
+            {
+                "run-data": {"prefix.run-data.2023-07": ["id1", "id2"]},
+                "run-toc": {
+                    "prefix.run-toc.2023-06": ["id3", "id4"],
+                    "prefix.run-toc.2023-07": ["id5", "id6"],
+                },
+                "funky-data": {"prefix.funky-data.1918-02": ["id1", "id2"]},
             },
-        }
-        map2 = {
-            "run-data": {"prefix.run-data.2023-06": ["id8", "id9"]},
-            "run-toc": {
-                "prefix.run-toc.2023-06": ["id10", "id12", "id13", "id14"],
-                "prefix.run-toc.2023-08": ["id20", "id21"],
+            {
+                "run-data": {"prefix.run-data.2023-06": ["id8", "id9"]},
+                "run-toc": {
+                    "prefix.run-toc.2023-06": ["id10", "id12", "id13", "id14"],
+                    "prefix.run-toc.2023-08": ["id20", "id21"],
+                },
+                "run-sample": {"prefix.run-sample.2023-08": ["id22", "id23"]},
             },
```

Comment:
I thought I replied to this, but maybe I made overlapping comment somewhere else:

1. The only place this code is actually used is in the tool data index pass, which is disabled and may never be enabled ... at least unless we radically rework it to be economic; I seriously (if only momentarily) debated just "bugchecking" if we ever got to that point in the indexer with an existing map. (Though that didn't feel right.)
2. We make *two* index passes (although the second is disabled), and the second pass can only have `tool-data-{tool}` indices in it, while the existing map (from the first pass) *can't* have those indices. So really, I could have reasonably ignored cases where the index root names overlapped, nevermind merging duplicate document IDs in identical indices. I didn't see any reason to deliberately omit root and index merging, and the alternatives would be to fail, skip, or somehow log -- at the time, writing and testing *that* code seemed to make even less sense.

So I guess I dug myself into a hole by coding, and testing, too much, and I just balked at going that extra imaginary and pointless step of worrying about the appearance on ID duplication.

1. The code has no sensitivity to document ID values: it just appends the two document ID lists, so if there's a duplicate it'll appear twice. The code doesn't care whether an ID repeats, so the case is already tested.
2. If we were to get there for real, the index is broken: Elasticsearch either complained about trying to overwrite an existing document or it quietly replaced the first document. In either case, the behavior of the merge itself isn't very interesting: that's something the indexing code would need to detect, long before we get here.

Reply:
So, before we dive into an interesting but otherwise possibly pointless philosophical discussion, the foremost reason for commenting on this snippet was to make sure that the code reads the way you intend it to.

I think that you added line 54 for a good and important reason (to ensure that there was a non-colliding (root) index in each map) but that you did so with a cut-and-paste of line 49, and you updated the indices but neglected the document IDs (which you've otherwise kept unique).  So, I wanted to make sure that you wanted the IDs to be duplicates...if so, we're all good, here.

OK, that said...

I think that there are two aspects to our situation.  We write the code to do at least enough to meet our requirements, and then we write tests to put the code through its paces.  I think it's reasonable for the tests to test what the code actually does or is capable of doing, even if the code turns out to be more capable than required.  This is in part because the demands on the code might change, especially if it is into an area which the code can already handle, and the tests won't necessarily be expanded at the time the demands increase.  If we were engaged in real, test-driven-design and practicing test-as-specification, then, sure, we probably would not produce the extra testing...but, in that situation, when the requirements expanded we would expand the testing, and we would have written the tests _first_.  Since that's not our actual practice, I think it's good to have the tests cover what the code _does_ (or what we expect it to do) rather than what we **need** (under the expectation that it does at least what we need).

So, yes, the requirement is to be able to merge a second map into a first, without any concern for document ID values; however, the code was written without any directionality, and it should produce the same result for the first merged into the second, and it would be good if it had a defined behavior regarding what it does with the document IDs, if only so that no one is surprised if it does that.

(As for Elasticsearch IDs, I think this map is used to generate bulk operations, and bulk operations are, I believe, more or less, executed serially on the server side, and I don't think ES cares if the same document is operated on more than once in a single bulk request...provided that the operations don't conflict with each other in some way (like changing the ownership twice would be OK, but deleting a document twice would likely lead to sadness).)

Reply:
> (As for Elasticsearch IDs, I think this map is used to generate bulk operations, and bulk operations are, I believe, more or less, executed serially on the server side, and I don't think ES cares if the same document is operated on more than once in a single bulk request...provided that the operations don't conflict with each other in some way (like changing the ownership twice would be OK, but deleting a document twice would likely lead to sadness).)

You're mostly right about that (although that refers to `stream` as opposed to `indices` which feeds the multi-index `_search` URI construction), but this misses the point about duplicate IDs. It's not concern over the behavior of the Elasticsearch bulk operation but rather the semantics of the index built by `pbench-index`. If we have document IDs from the two indexing passes they *aren't* the same document, so the fact that we generated the same document ID twice is a concern. It's not an index mapping problem, it's an *indexing* problem. That is, the damage has been done and if the indexer failed to notice, and let us get this far, there's nothing we can do about it. And, again, it's also pretty hypothetical as the two passes aren't intended to touch the same root index names and the likelihood of generating identical documents in the two passes is effectively nil. 

And we really *do* already test what would happen, by exercising the merging of document lists across an identical index name in the two passes: nothing cares about the value of the document IDs, so there's no special case. And, so far as black box semantics goes, we don't care how it handles duplicates as I have no interest in placing (much less testing) imaginary interface constraints for a "don't care / can't happen" thought experiment.